File: javascript/src/utils/scoundrel-json.js
// @ts-check

const TYPE_KEY = "__scoundrel_type__"
const VALUE_KEY = "value"
const OMIT_VALUE = Symbol("omit_value")

/** @typedef {{type: string, canSerialize: (value: any) => boolean, serialize: (value: any) => Record<string, any>, deserialize: (payload: Record<string, any>) => any}} ScoundrelTypeHandler */

/** @type {ScoundrelTypeHandler[]} */
const typeHandlers = []

const isPlainObject = (value) => {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false
  const prototype = Object.getPrototypeOf(value)
  return prototype === Object.prototype || prototype === null
}

const findHandlerForValue = (value) => typeHandlers.find((handler) => handler.canSerialize(value))
const findHandlerForType = (type) => typeHandlers.find((handler) => handler.type === type)

/**
 * Registers a new scoundrel type handler.
 * @param {ScoundrelTypeHandler} handler Handler to register
 */
export function registerScoundrelType(handler) {
  if (!handler || typeof handler.type !== "string") {
    throw new Error("Scoundrel type handler must include a type string")
  }
  if (typeof handler.canSerialize !== "function" || typeof handler.serialize !== "function" || typeof handler.deserialize !== "function") {
    throw new Error(`Scoundrel type handler '${handler.type}' must define canSerialize, serialize, and deserialize`)
  }

  const existingIndex = typeHandlers.findIndex((existing) => existing.type === handler.type)
  if (existingIndex >= 0) {
    typeHandlers[existingIndex] = handler
  } else {
    typeHandlers.push(handler)
  }
}

const ensureSerializableObject = (handler, serialized, path) => {
  if (!isPlainObject(serialized)) {
    throw new Error(`Scoundrel type '${handler.type}' must serialize to a plain object at ${path}`)
  }

  if (!Object.prototype.hasOwnProperty.call(serialized, TYPE_KEY)) {
    serialized[TYPE_KEY] = handler.type
  }

  return serialized
}

const encodeValue = (value, path, seen) => {
  if (typeof value === "undefined") {
    return path === "value" ? null : OMIT_VALUE
  }

  const handler = findHandlerForValue(value)
  if (handler) {
    const serialized = ensureSerializableObject(handler, handler.serialize(value), path)
    return encodeObject(serialized, path, seen)
  }

  const valueType = typeof value

  if (valueType === "function") throw new Error(`Cannot serialize function at ${path}`)
  if (valueType === "symbol") throw new Error(`Cannot serialize symbol at ${path}`)
  if (valueType === "number" && !Number.isFinite(value)) {
    throw new Error(`Cannot serialize non-finite number at ${path}`)
  }
  if (valueType === "bigint") {
    const numberValue = Number(value)
    if (!Number.isFinite(numberValue)) {
      throw new Error(`Cannot serialize non-finite number at ${path}`)
    }
    return numberValue
  }

  if (value === null || valueType !== "object") return value

  if (Array.isArray(value)) {
    if (seen.has(value)) throw new Error(`Cannot serialize circular reference at ${path}`)
    seen.set(value, path)
    return value.map((item, index) => {
      const encoded = encodeValue(item, `${path}[${index}]`, seen)
      return encoded === OMIT_VALUE ? null : encoded
    })
  }

  if (!isPlainObject(value)) {
    const constructorName = value.constructor && value.constructor.name ? value.constructor.name : "Object"
    throw new Error(`Cannot serialize non-plain object '${constructorName}' at ${path}`)
  }

  return encodeObject(value, path, seen)
}

const encodeObject = (value, path, seen) => {
  if (seen.has(value)) throw new Error(`Cannot serialize circular reference at ${path}`)
  seen.set(value, path)

  /** @type {Record<string, any>} */
  const encoded = {}

  for (const key of Object.keys(value)) {
    const childPath = `${path}.${key}`
    const childValue = encodeValue(value[key], childPath, seen)
    if (childValue !== OMIT_VALUE) {
      encoded[key] = childValue
    }
  }

  return encoded
}

const decodeValue = (value) => {
  if (Array.isArray(value)) {
    return value.map((item) => decodeValue(item))
  }

  if (value && typeof value === "object") {
    const type = value[TYPE_KEY]
    if (typeof type === "string") {
      const handler = findHandlerForType(type)
      if (handler) {
        return handler.deserialize(value)
      }
    }

    /** @type {Record<string, any>} */
    const decoded = {}
    for (const key of Object.keys(value)) {
      decoded[key] = decodeValue(value[key])
    }
    return decoded
  }

  return value
}

/**
 * Serializes a value using Scoundrel JSON rules.
 * @param {any} value Value to serialize
 * @returns {string} JSON string representation
 */
export function serializeScoundrelJSON(value) {
  const encoded = encodeValue(value, "value", new WeakMap())
  return JSON.stringify(encoded)
}

/**
 * Parses a JSON payload using Scoundrel JSON rules.
 * @param {string} raw JSON string payload
 * @returns {any} Parsed value
 */
export function parseScoundrelJSON(raw) {
  return decodeValue(JSON.parse(raw))
}

registerScoundrelType({
  type: "date",
  canSerialize: (value) => value instanceof Date,
  serialize: (value) => ({[TYPE_KEY]: "date", [VALUE_KEY]: value.toISOString()}),
  deserialize: (payload) => new Date(payload[VALUE_KEY])
})

registerScoundrelType({
  type: "regex",
  canSerialize: (value) => value instanceof RegExp,
  serialize: (value) => ({[TYPE_KEY]: "regex", [VALUE_KEY]: value.toString()}),
  deserialize: (payload) => {
    const raw = payload[VALUE_KEY]
    if (typeof raw !== "string") throw new Error("Invalid regex payload")
    const match = raw.match(/^\/(.*)\/([a-z]*)$/i)
    if (!match) throw new Error("Invalid regex payload")
    return new RegExp(match[1], match[2])
  }
})

File: python/scoundrel_python/scoundrel_json.py
from __future__ import annotations

import datetime as dt
import json
import math
import re
from dataclasses import dataclass
from typing import Any, Callable, Dict, List

TYPE_KEY = "__scoundrel_type__"
VALUE_KEY = "value"


@dataclass(frozen=True)
class ScoundrelTypeHandler:
  type: str
  can_serialize: Callable[[Any], bool]
  serialize: Callable[[Any], Dict[str, Any]]
  deserialize: Callable[[Dict[str, Any]], Any]


_type_handlers: List[ScoundrelTypeHandler] = []


def register_scoundrel_type(handler: ScoundrelTypeHandler) -> None:
  if not handler.type:
    raise ValueError("Scoundrel type handler must include a type")

  existing_index = next((idx for idx, item in enumerate(_type_handlers) if item.type == handler.type), None)
  if existing_index is None:
    _type_handlers.append(handler)
  else:
    _type_handlers[existing_index] = handler


def _find_handler_for_value(value: Any) -> ScoundrelTypeHandler | None:
  return next((handler for handler in _type_handlers if handler.can_serialize(value)), None)


def _find_handler_for_type(type_name: str) -> ScoundrelTypeHandler | None:
  return next((handler for handler in _type_handlers if handler.type == type_name), None)


def _ensure_serialized_object(handler: ScoundrelTypeHandler, payload: Any, path: str) -> Dict[str, Any]:
  if not isinstance(payload, dict):
    raise TypeError(f"Scoundrel type '{handler.type}' must serialize to a dict at {path}")

  if TYPE_KEY not in payload:
    payload[TYPE_KEY] = handler.type

  return payload


def _encode_value(value: Any, path: str, seen: Dict[int, str]) -> Any:
  if value is None or isinstance(value, (str, bool, int)):
    return value

  if isinstance(value, float):
    if not math.isfinite(value):
      raise ValueError(f"Cannot serialize non-finite number at {path}")
    return value

  handler = _find_handler_for_value(value)
  if handler is not None:
    serialized = _ensure_serialized_object(handler, handler.serialize(value), path)
    return _encode_object(serialized, path, seen)

  if isinstance(value, (list, tuple)):
    if id(value) in seen:
      raise ValueError(f"Cannot serialize circular reference at {path}")
    seen[id(value)] = path
    return [_encode_value(item, f"{path}[{index}]", seen) for index, item in enumerate(value)]

  if isinstance(value, dict):
    return _encode_object(value, path, seen)

  raise TypeError(f"Cannot serialize unsupported type {type(value).__name__} at {path}")


def _encode_object(value: Dict[str, Any], path: str, seen: Dict[int, str]) -> Dict[str, Any]:
  if id(value) in seen:
    raise ValueError(f"Cannot serialize circular reference at {path}")
  seen[id(value)] = path

  encoded: Dict[str, Any] = {}
  for key, child in value.items():
    key_str = str(key)
    child_path = f"{path}.{key_str}"
    encoded[key_str] = _encode_value(child, child_path, seen)
  return encoded


def _decode_value(value: Any) -> Any:
  if isinstance(value, list):
    return [_decode_value(item) for item in value]

  if isinstance(value, dict):
    type_name = value.get(TYPE_KEY)
    if isinstance(type_name, str):
      handler = _find_handler_for_type(type_name)
      if handler is not None:
        return handler.deserialize(value)
    return {key: _decode_value(child) for key, child in value.items()}

  return value


def dumps(value: Any) -> str:
  encoded = _encode_value(value, "value", {})
  return json.dumps(encoded)


def loads(raw: str) -> Any:
  return _decode_value(json.loads(raw))


def _parse_iso_datetime(raw: str) -> dt.datetime:
  if raw.endswith("Z"):
    raw = raw[:-1] + "+00:00"
  return dt.datetime.fromisoformat(raw)


def _format_datetime(value: dt.datetime) -> str:
  if value.tzinfo is None:
    return value.isoformat()
  return value.astimezone(dt.timezone.utc).isoformat().replace("+00:00", "Z")


def _regex_flags_from_pattern(pattern: re.Pattern) -> str:
  flags = ""
  if pattern.flags & re.IGNORECASE:
    flags += "i"
  if pattern.flags & re.MULTILINE:
    flags += "m"
  if pattern.flags & re.DOTALL:
    flags += "s"
  if pattern.flags & re.UNICODE:
    flags += "u"
  return flags


def _regex_flags_to_re(flags: str) -> int:
  flag_value = 0
  if "i" in flags:
    flag_value |= re.IGNORECASE
  if "m" in flags:
    flag_value |= re.MULTILINE
  if "s" in flags:
    flag_value |= re.DOTALL
  if "u" in flags:
    flag_value |= re.UNICODE
  return flag_value


register_scoundrel_type(
  ScoundrelTypeHandler(
    type="date",
    can_serialize=lambda value: isinstance(value, dt.datetime),
    serialize=lambda value: {TYPE_KEY: "date", VALUE_KEY: _format_datetime(value)},
    deserialize=lambda payload: _parse_iso_datetime(str(payload[VALUE_KEY]))
  )
)

register_scoundrel_type(
  ScoundrelTypeHandler(
    type="regex",
    can_serialize=lambda value: isinstance(value, re.Pattern),
    serialize=lambda value: {TYPE_KEY: "regex", VALUE_KEY: f"/{value.pattern}/{_regex_flags_from_pattern(value)}"},
    deserialize=lambda payload: _deserialize_regex(payload)
  )
)


def _deserialize_regex(payload: Dict[str, Any]) -> re.Pattern:
  raw = str(payload[VALUE_KEY])
  match = re.match(r"^/(.*)/([a-z]*)$", raw)
  if not match:
    raise ValueError("Invalid regex payload")
  pattern, flags = match.groups()
  return re.compile(pattern, flags=_regex_flags_to_re(flags))
